include/diy/master.hpp

```diff
@@ -1008,21 +1024,30 @@ comm_exchange(ToSendList& to_send, int out_queues_limit)
     if (ostatus->tag() == tags::queue)
     {
       size_t size  = ir.message.size();
+      int from = ir.info.from;
+      int to = ir.info.to;
       int external = -1;
 
-      if (block(lid(to)) != 0 || !queue_policy_->unload_incoming(*this, from, to, size))
+      IncomingRound *in = &incoming_;
+      if (ir.info.round != exchange_round_)
       {
-        incoming_[to].queues[from].swap(ir.message);
-        incoming_[to].queues[from].reset();     // buffer position = 0
+        assert(ir.info.round > exchange_round_);
+        in = &incoming_next_[ir.info.round];
       }
-      else if (queue_policy_->unload_incoming(*this, from, to, size))
+
+      if (block(lid(to)) == 0 && queue_policy_->unload_incoming(*this, from, to, size))
```

Comment:
I think here you have to treat the case of the queues from future rounds specially. It's possible that the block is in memory, but will be unloaded in the future. Then the queue from future round should not stay in memory, I don't think.